alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub(crate) struct ChildProcessState {
+    _on_exit: HandleWaitSignal,
+    events: Receiver<ChildEvent>,
+}
+
+impl ChildProcessState {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessState, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            if let Err(e) = sender.send(ChildEvent::Exited) {
+                warn!(
+                    "An error occurred while attempting to notify about child process z
+                    termination: {}",
+                    e
+                );
+            }
+        })?;
+
+        Ok(ChildProcessState { _on_exit: on_exit, events: receiver })
+    }
+
+    pub fn events(&self) -> &Receiver<ChildEvent> {
+        &self.events
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Error;
+    use std::ptr;
+    use std::time::Duration;
+
+    use std::sync::mpsc::channel;
+
+    use widestring::U16CString;
+
+    use winapi::shared::ntdef::LPWSTR;
+    use winapi::um::processthreadsapi::{
+        CreateProcessW, TerminateProcess, PROCESS_INFORMATION, STARTUPINFOW,
+    };
+
+    use super::*;
+
+    fn make_cmd_process() -> Result<HANDLE, Error> {
+        let mut pi: PROCESS_INFORMATION = Default::default();
+        let mut si: STARTUPINFOW = Default::default();
+        let cmdline = U16CString::from_str("cmd.exe").unwrap();
+
+        unsafe {
+            if 0 == CreateProcessW(
+                ptr::null(),
+                cmdline.as_ptr() as LPWSTR,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                false as i32,
+                Default::default(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                &mut si as *mut STARTUPINFOW,
+                &mut pi as *mut PROCESS_INFORMATION,
+            ) {
+                Err(Error::last_os_error())
+            } else {
+                Ok(pi.hProcess)
+            }
+        }
+    }
+
+    // TODO: This is more of an integration test since it spawns new 'cmd.exe'
+    //       process and heavily uses Win32 API. Should this be ignored
+    //       by default? Worst-case, this will timeout after `WAIT_TIMEOUT`.
+    #[test]
+    pub fn on_handle_wait_completed_signalls() {
```

Comment:
```suggestion
    pub fn on_handle_wait_completed_signals() {
```

Reply:
Renamed it to `on_handle_wait_runs_callback_when_process_exits` in 74b7b9a